### PR TITLE
Upgrade Kotlin to 2.0.0

### PR DESCRIPTION
**Pull request by vegaro**
- [x] Upgrade Kotlin to 2.0.0
- [ ] Check hybrids keep working
  - [ ] Flutter
  - [x] Unity
- [x] Upgrade Compose to compatible version
- [x] Check Android manifest differences (compared purchase tester generated AndroidManifest.xml before and after)
- [ ] Check full dependency tree changes
- [x] Dokka keeps working
- [ ] Check that clients on older version of compose can still use paywalls

### CODE REVIEW

File: examples/paywall-tester/build.gradle [hunk 1]
```diff
@@ -45,17 +46,11 @@ android {
         targetCompatibility JavaVersion.VERSION_1_8
     }
     kotlinOptions {
-        jvmTarget = '1.8'
-    }
-    buildFeatures {
-        compose true
+        jvmTarget = '19'
```
Comment (JayShortway) [inline examples/paywall-tester/build.gradle:49]: This seems quite high and doesn't correspond to the `targetCompatibility` set above. Is this needed for anything? (Also in `api-tester/build.gradle`.)

File: build.gradle [hunk 2]
```diff
@@ -16,6 +16,7 @@ plugins {
 //    The request for this plugin could not be satisfied because the plugin is already on the classpath with an
 //    unknown version, so compatibility cannot be checked.
     alias libs.plugins.kotlin.parcelize apply false
+    alias libs.plugins.compose.compiler apply false
```
Comment (JayShortway) [inline build.gradle:19]: Nice improvement!

File: gradle/libs.versions.toml [hunk 3]
```diff
@@ -109,9 +110,9 @@ robolectric = { module = "org.robolectric:robolectric", version.ref = "roboelect
 
 tink = { module = "com.google.crypto.tink:tink-android", version.ref = "tink" }
 
-coroutines-test = { module = "org.jetbrains.kotlinx:kotlinx-coroutines-test", version.ref = "coroutines" }
+coroutines-test = { module = "org.jetbrains.kotlinx:kotlinx-coroutines-test", version = "1.8.1" }
 
-kotlin-bom = "org.jetbrains.kotlin:kotlin-bom:1.8.0"
+kotlin-bom = "org.jetbrains.kotlin:kotlin-bom:2.0.0"
```
Comment (JayShortway) [inline gradle/libs.versions.toml:115]:
```suggestion
kotlin-bom = { module = "org.jetbrains.kotlin:kotlin-bom", version.ref = "kotlin" }
```

**Comment by JayShortway** [comment]
Is my assumption correct that updating to Kotlin 2.0.0 means implementing apps will need to at least be on 1.8.0?

**Comment by JayShortway** [comment]
Looks good, nice update! Just a few minor things.

File: examples/paywall-tester/build.gradle (hunk 1 above)
Comment (vegaro) [reply examples/paywall-tester/build.gradle:49]: hmm, good point, I saw it in a snippet in https://developer.android.com/jetpack/androidx/releases/compose-kotlin#declaring_dependencies 

what do you think? if I understand correctly, this is the version of the jvm our code should be compatible with? So maybe we should keep it as 1.8?

**Comment by vegaro** [comment]
> Is my assumption correct that updating to Kotlin 2.0.0 means implementing apps will need to at least be on 1.8.0?

yes

File: examples/paywall-tester/build.gradle (hunk 1 above)
Comment (JayShortway) [reply examples/paywall-tester/build.gradle:49]: Yeah that's how I understand it as well. [Java 17 is the latest supported version on Android 14](https://developer.android.com/build/jdks#compileSdk), and [desugaring is only available for Java 11 & 8](https://developer.android.com/studio/write/java8-support#library-desugaring). I have no idea why that snippet sets it to 19. 😕  
  
I think keeping it at 1.8 (8) is the safest indeed, unless we need any of the new Java features.

**Comment by tonidero** [comment]
Looks good so far! We should also check that clients on older version of compose can still use paywalls (or at least to figure out if there is a min version)

**Comment by JayShortway** [comment]
Just came across [this](https://kotlinlang.org/docs/whatsnew20.html#new-directory-for-kotlin-data-in-gradle-projects), might be good to add a line to `.gitignore`? 

> With this change, you may need to add the `.kotlin` directory to your project's `.gitignore` file.

**Comment by vegaro** [comment]
alright, after a bunch of testing

Looks like we can only upgrade kotlin to 1.8 and 2.0 even 1.9 are going to have to wait. The main reasoning is:
- Capacitor uses kotlin 1.8
- React Native latest version (0.74) uses 1.9, but that’s very recent
- React Native 0.72 uses 1.7, and I’ve managed to build an app in this version if we upgrade to 1.8. If we upgrade to 1.9 it starts breaking. I also tried with 0.71 (using 1.6) and it also worked . If anyone has issues, we at least know that 0.73 should work fine, so we can recommend to upgrade.

I will close the 2.0 PR and open one upgrading to 1.8

**Pull request closed without merging**